tests/test_providers.py

```diff
@@ -0,0 +1,125 @@
+from typing import Dict, Type, Any
+
+import attr
+from pytest_twisted import inlineCallbacks
+from twisted.python.failure import Failure
+
+import scrapy
+from scrapy import Request
+from scrapy.crawler import Crawler
+from scrapy.settings import Settings
+from scrapy_poet.page_input_providers import PageObjectInputProvider
+from tests.utils import crawl_single_item, HtmlResource
+
+
+class ProductHtml(HtmlResource):
+    html = """
+    <html>
+        <div class="breadcrumbs">
+            <a href="/food">Food</a> / 
+            <a href="/food/sweets">Sweets</a>
+        </div>
+        <h1 class="name">Chocolate</h1>
+        <p>Price: <span class="price">22€</span></p>
+        <p class="description">The best chocolate ever</p>
+    </html>
+    """
+
+
+@attr.s(auto_attribs=True)
+class Price:
+    price: str
+
+
+@attr.s(auto_attribs=True)
+class Name:
+    name: str
+
+
+@attr.s(auto_attribs=True)
+class Html:
+    html: str
+
+
+class PriceHtmlDataProvider(PageObjectInputProvider):
+
+    provided_classes = {Price, Html}
+
+    def __init__(self, crawler: Crawler):
+        assert isinstance(crawler, Crawler)
+        super().__init__(crawler)
+
+    def __call__(self, to_provide, response: scrapy.http.Response, spider: scrapy.Spider):
+        assert isinstance(spider, scrapy.Spider)
+        ret: Dict[Type, Any] = {}
+        if Price in to_provide:
+            ret[Price] = response.css(".price::text").get()
```

Comment:
hm, what's the use case of allowing to return a value of a non-matching type? Here for Price a str is returned. 

A related question: why do users need to write `return {cls: instance_of_cls}`, instead of `return [instance_of_cls]` - is it because of some use cases for returning not exactly the same type? What are these use cases? Sorry if I misssed that in the previous discussions.

Reply:
I tried to trace back this decision. I found this reference: https://github.com/scrapinghub/scrapy-poet/pull/40#discussion_r491875341 that stated that this is still an open question. So let me philosophize about it. 

We have only two possibilities IMO:
1) Keep the current structure (but include a check that makes sure that the returned object is a subclass of the expected type for security). 
2) Force to always return instances of the expected type. In this case the return `dict` could be replaced by a `list`. 

Option 2. clearly makes the API cleaner, but disallow supporting instances that are a subclass of the expected type. So the question we have to answer is: **is the functionality of allowing returning subclasses of the expected type useful**. I'll focus on listing some examples where it could be useful. This is the full list of cases:

```py
[]
```

I can think of cases where subclassing the provider to modify its behaviour makes sense (e.g. subclassing the default `ReponseDataProvider` to apply ad-block rules to the HTML), but the returning instance would have the same type: it is enough with changing its data. 

But I don't see a clear case where the returning instance requires a change in its business logic, especially given that these objects are usually just data carriers. So I'm going to go for option 2 and simplify the API. @kmike what do you think?

Reply:
This sounds good @ivanprado. I also can't think of a use case. 

Even if we find a use case in future, we'd be able to adjust API in a backwards compatible way, but we won't need to pay the complexity price in the meantime.

Reply:
Providers returning a list of instances implemente here: https://github.com/scrapinghub/scrapy-poet/pull/41/commits/24b9627ff4ff0749c9f24bbf5ab6bc6ff37b88af